### DISCUSSION TITLE
How to use TTF font file?

### DISCUSSION THREAD

**Discussion by ohjohnsen** [AvaloniaUI/Avalonia#10302] [Q&A] [answered]
I'm trying to use a custom monospace TTF file as a font in my application.  I've downloaded a TTF file from Google Fonts (Roboto Mono Regular).  But I can't figure out how to load the font file in my application.  I've searched this for a couple of hours now, but I can't find any documentation on it either.

How can this be done?  I have the file in a Resources/Fonts directory in my project.

**Comment by maxkatz6** [comment]
Fonts are assets. So you need to include them as an AvaloniaResource. See https://docs.avaloniaui.net/docs/getting-started/assets#including-assets for details.

And then you can use it from the XAML as in "Referencing assets" paragraph of the same page.

**Reply by maxkatz6** [reply]
In a similar way you can define font in application/control resources:
```xaml
<Application.Resources>
  <FontFamily x:Key="ApplicationFont">avares://MyAssembly/Assets/font.ttf#FontName</FontFamily>
</Application.Resources>
```

and then used with StaticResource:
```xaml
<Window FontFamily="{StaticResource ApplicationFont}" />
```

**Reply by maxkatz6** [reply]
In your case, you can include fonts into the project something like:
```xml
<AvaloniaResource Include="Resources\Fonts\**.*" \>
```

**Reply by ohjohnsen** [reply]
Thank you for your answer. But I can't seem to get my head around this.  To summarise my understanding of what you write:

1. Since the project file already has "Assets\Fonts" as a folder to include, I renamed my folder in the project from "Resources" to "Assets".  So the path to the TTF file is currently "SerialToolMk2\SerialToolMk2\Assets\Fonts\RobotoMono-Regular.ttf" (both solution and project is named "SerialToolMk2").
2. In App.axaml I've added the TTF file as an application resource:
```
<Application.Resources>
    <FontFamily x:Key="ApplicationFont">
        avares://SerialToolMk2/Assets/Fonts/RobotoMono-Regular.ttf#RobotoMonoRegularFontFamily
    </FontFamily>
</Application.Resources>
```

3.  In the design I use the font as a static resource:
```
<TextBlock
    Text="{Binding String}"
    FontFamily="{StaticResource RobotoMonoRegularFontFamily}" />                         
```

Then, when building and running the application, it crashes when rendering the window, with the following error message:
"Unhandled exception. System.Collections.Generic.KeyNotFoundException: Static resource 'RobotoMonoRegularFontFamily' not found."

Can you see what I'm doing wrong here?

**Reply by maxkatz6** [reply]
Aside from real font name, you defined resource with a one key, but used with another in a static resource.

**Reply by ohjohnsen** [reply]
Bah.. Ok, that was stupid. 🙄 Setting the Key value fixes the error coming when using the font as a static resource.  But I still get an error on the font name:
```
<Application.Resources>
    <FontFamily x:Key="RobotoMonoFontFamily">avares://SerialToolMk2/Assets/Fonts/RobotoMono-Regular.ttf#Roboto Mono</FontFamily>
</Application.Resources>
```

`Unhandled exception. System.InvalidOperationException: Could not create glyph typeface for: Roboto Mono.`

**Reply by Enigma205** [reply]
You have to use the Key not the Font name after the # sign.

```
<TextBlock
    Text="{Binding String}"
    FontFamily="{StaticResource ApplicationFont}" />  
```

**Comment by Gillibald** [comment]
You specify the font family without its real name. The part after # represents the name. Here it is probably #Roboto Mono

**Reply by ohjohnsen** [reply]
Thank you for your answer.  Should this be correct?
```
<Application.Resources>
    <FontFamily x:Key="RobotoMonoFontFamily">avares://SerialToolMk2/Assets/Fonts/RobotoMono-Regular.ttf#Roboto Mono</FontFamily>
</Application.Resources>
```

I've tried various things as font name: `Roboto`, `Roboto Mono`, `Roboto Mono Regular`.  But in any case I get an error thrown:

`Unhandled exception. System.InvalidOperationException: Could not create glyph typeface for: Roboto Mono.`

**Comment by Gillibald** [comment, marked as answer]
Please share your csproj file

**Reply by ohjohnsen** [reply]
Yes, here's the csproj file: https://gist.github.com/ohjohnsen/9eaa3b06fd39ab74ca4491ea23589516

I also saw in the documentation that the resource could (should?) be embedded in the assembly DLL file, and be referenced to using the "avares" path (https://docs.avaloniaui.net/docs/getting-started/assets#referencing-assets).  But I notice that my assembly DLL file is the same file no matter if the resource is defined in `App.axaml` or not..  Could this be the problem?

**Reply by Gillibald** [reply]
Add this ` <AvaloniaResource Include="Assets\Fonts\*" />` to your csproj

**Reply by ohjohnsen** [reply]
Perfect!  I added a new PropertyGroup, with the AvaloniaResource tag inside.  And now the assembly DLL file is larger (evidently including the font file), and the font is loaded and used.

Thanks a lot for your assistance.

**Comment by arshamX** [comment]
some how i have this issue too i have added my font to my cross platform application it works on desktop but it doesnt work in android
